File: scala-project/src/main/scala/au/id/tmm/presentations/numbertypes/SignedIntegers.scala
package au.id.tmm.presentations.numbertypes

object SignedIntegers {

  def main(args: Array[String]): Unit = {
    overflows()
  }

  private def range(): Unit = {
    println(Int.MinValue)
    println(Int.MaxValue)
  }

  private def overflows(): Unit = {
    println(Int.MaxValue + 1)
    println(0 - 1)
  }

  private def safeArithmetic(): Unit = {
    println(math.addExact(1, 1))
    println(math.addExact(Int.MaxValue, 1))
  }

  private def division(): Unit = {
    println(2 / 2)
    println(2 / 3)
  }

  private def divisionByZero(): Unit = {
    println(1 / 0)
  }

  private def zeroDivisionByZero(): Unit = {
    println(0 / 0)
  }

}

File: kotlin/src/main/kotlin/au/id/tmm/presentations/numbertypes2/SignedIntegers.kt
package au.id.tmm.presentations.numbertypes2

object SignedIntegers {

  @JvmStatic
  fun main(args: Array<String>) {
    // range()
    // overflows()
    // safeArithmetic()
    // division()
    // divisionByZero()
    // divisionZeroByZero()
  }

  private fun range() {
    println(Int.MIN_VALUE)
    println(Int.MAX_VALUE)
  }

  private fun overflows() {
    println(Int.MAX_VALUE + 1)
    println(0 - 1)
  }

  private fun safeArithmetic() {
    println(StrictMath.addExact(1, 1))
    println(StrictMath.addExact(Int.MAX_VALUE, 1))
  }

  private fun division() {
    println(2 / 2)
    println(2 / 3)
  }

  private fun divisionByZero() {
    println(1 / 0)
  }

  private fun divisionZeroByZero() {
    println(0 / 0)
  }

}